src/Sarif.UnitTests/TestData/SarifVersionOneToCurrentVisitor/ExpectedOutputs/OneRunWithRules.sarif

```diff
@@ -68,14 +71,13 @@
               "text": "Rent internal rebellion competence biography photograph."
             },
             "configuration": {
-              "defaultLevel": "note"
+              "defaultLevel": "note",
+              "defaultRank": 0.0
             }
-          },
-          "C2002-1": {
```

Comment:
>"C2002-1" [](start = 10, length = 9)

The v1 file has a `result` with a `ruleKey` that does not exist in the `rules` dictionary. The visitor code was constructing an empty `rule` object for that key to refer to. I don't see the point of that. The new code just sets the v2 `ruleIndex` to -1 in this case. See how on the right-hand side of this diff, the `result` starting on L28 has no `ruleIndex`. The old code would have given it a synthetic one.

Reply:
ok. side issue: what does v2 say about no rule id and ruleIndex == -1. Seems like an invalid condition. A rule (especially for a converted log file) might not have the SARIF notion of a rule id but in this case, should have a name, so the rule object should exist.

IOW, suggest we say in the spec that either rule id or ruleIndex (with a corresponding rule object) SHOULD be present.

---
In reply to: [250286632](https://github.com/Microsoft/sarif-sdk/pull/1232#discussion_r250286632) [](ancestors = 250286632)

Reply:
We discussed this in person, but for the record: This condition is valid. If a tool does not provide opaque, stable rule ids as required by the spec, `ruleId` must be absent. If the log file does not provide rule metadata (or at least, does not provide metadata for the rule implicated by this result), then `ruleIndex` must be absent/-1.

---
In reply to: [250389698](https://github.com/Microsoft/sarif-sdk/pull/1232#discussion_r250389698) [](ancestors = 250389698,250286632)